src/modules/navigator/mission_reverse.cpp

```diff
@@ -0,0 +1,284 @@
+	if ((_mission_item.nav_cmd == NAV_CMD_LAND)
+	    && (_navigator->abort_landing())) {
+
+		_mission->do_abort_landing();
+	}
+}
+
+void
+MissionReverse::switch_from_nonmission()
+{
+	_switch_from_nonmission = true;
+}
+
+void
+MissionReverse::advance_mission()
+{
+	dm_item_t dm_current = (dm_item_t)(_mission->_offboard_mission.dataman_id);
+
+	for (int32_t i = _mission->_current_offboard_mission_index - 1; i >= 0; i--) {
+		struct mission_item_s missionitem = {};
+		const ssize_t len = sizeof(missionitem);
+
+		if (dm_read(dm_current, i, &missionitem, len) != len) {
+			/* not supposed to happen unless the datamanager can't access the SD card, etc. */
+			PX4_ERR("dataman read failure");
+			break;
+		}
+
+		if (item_contains_position(missionitem)) {
+			_mission->_current_offboard_mission_index = i;
+			return;
+		}
+	}
+
+	// finished flying back the mission
+	_mission->_current_offboard_mission_index = -1;
+}
+
+void
+MissionReverse::set_mission_items()
+{
+	_mission->set_mission_items(true);
+	_mission_reverse_finished = _mission->_mission_type == Mission::MISSION_TYPE_NONE;
+}
+
+
+uint16_t
+MissionReverse::index_closest_mission_item() const
+{
+	uint16_t min_dist_index = _mission->index_closest_mission_item();
+	float dist_xy(FLT_MAX), dist_z(FLT_MAX);
+
+	dm_item_t dm_current = (dm_item_t)(_mission->_offboard_mission.dataman_id);
+	struct mission_item_s missionitem = {};
+	const ssize_t len = sizeof(missionitem);
+
+	if (dm_read(dm_current, min_dist_index, &missionitem, len) != len) {
+		/* not supposed to happen unless the datamanager can't access the SD card, etc. */
+		PX4_ERR("dataman read failure");
+	}
+
+	float min_dist = get_distance_to_point_global_wgs84(missionitem.lat, missionitem.lon,
+			 get_absolute_altitude_for_item(missionitem),
+			 _navigator->get_global_position()->lat,
+			 _navigator->get_global_position()->lon,
+			 _navigator->get_global_position()->alt,
+			 &dist_xy, &dist_z);
+
+	// also consider the home position
+	float dist = get_distance_to_point_global_wgs84(
+			     _navigator->get_home_position()->lat,
+			     _navigator->get_home_position()->lon,
+			     _navigator->get_home_position()->alt,
+			     _navigator->get_global_position()->lat,
+			     _navigator->get_global_position()->lon,
+			     _navigator->get_global_position()->alt,
+			     &dist_xy, &dist_z);
+
+	if (dist < min_dist) {
+		min_dist = dist;
+		min_dist_index = -1;
+	}
+
+	return min_dist_index;
+}
+
+bool
+MissionReverse::mission_changed()
+{
+	if ((_previous_mission.timestamp == _mission->_offboard_mission.timestamp) &&
+	    (_previous_mission.current_seq == _mission->_offboard_mission.current_seq) &&
+	    (_previous_mission.count == _mission->_offboard_mission.count) &&
+	    (_previous_mission.dataman_id == _mission->_offboard_mission.dataman_id)) {
+		return false;
+
+	} else {
+		_previous_mission = _mission->_offboard_mission;
+		return true;
+	}
+}
+
+void
+MissionReverse::command_vtol_transition()
```

Comment:
This should probably move to MissionBlock as well with a parameter for which transition. It's needed in several places (RTL, Land, Mission, etc).

Reply:
It is differently used in the Mission and Land so I am not sure if it is really helpful to generalize it to MissionBlock.

Reply:
I don't see why it shouldn't all be the same. When you use the mission item in a MissionBlock to do the transition we can also wait for acceptance before continuing.

https://github.com/PX4/Firmware/blob/master/src/modules/navigator/mission.cpp#L610-L612
https://github.com/PX4/Firmware/blob/master/src/modules/navigator/mission.cpp#L663-L665

Reply:
Ok, I initially thought of putting the full function to MissionBlock. I now created function which sets the correct values for the mission_item as this is the same for all different use cases.